jsonComposer.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2018-2019 Couchbase, Inc. All rights reserved.
+
+package gojsonsm
+
+import (
+	"fmt"
+)
+
+type jsonObjComposer struct {
+	// should have enough length
+	body []byte
+	// cursor
+	pos int
+	// if the newly composed json has atleast one field
+	atLeastOneFieldLeft bool
+}
+
+func (composer *jsonObjComposer) Write(data []byte) {
+	copy(composer.body[composer.pos:composer.pos+len(data)], data)
+	composer.pos += len(data)
+}
+
+func (composer *jsonObjComposer) Commit(stepBackAndClose bool) (int, bool) {
+	// close the JSON object after stepping back by one position explicity if requested
+	if stepBackAndClose && composer.pos-1 >= 0 {
+		if composer.body[composer.pos-1] != '{' {
```

Comment:
stepBackAndClose is for when we would have added a , to the composer and we have to step back, and overwrite it with a } to close the JSON correctly, when we removed the last item of the object. Eg: `{"key":"value","key1":"value1"}` on removal of "key1" becomes `{"key":"value",` - now to close it, we will have to step back and overwrite , with }.

But consider a simplest case of src = `{"key":"value"}` and to remove "key"
token { - write
"key":"value" - don't write
stepBackAndClose will be true here because it is the last item that we removed.
But in this case we can't step back and add a }, because there is no `,` to overwrite, but we will end up overriding the `{` that was written. It is to tackle this edge case of the last item and first item removed is the same.

Reply:
This is now moved as part of `Write()` to support any depth key removal.